sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -3878,6 +3880,8 @@ class Dataset[T] private[sql](
    */
   def persist(newLevel: StorageLevel): this.type = {
     sparkSession.sharedState.cacheManager.cacheQuery(this, None, newLevel)
+    queryExecution = new QueryExecution(
+      queryExecution.sparkSession, queryExecution.logical, queryExecution.tracker)
```

Comment:
there is also `mode: CommandExecutionMode.Value` argument - I don't know its meaning, but I'm pretty sure it'd need to be set too